presto-main/src/main/java/com/facebook/presto/operator/HashPartitionMaskOperator.java

```diff
@@ -220,12 +220,13 @@ public void addInput(Page page)
         }
 
         // build output page
-        Block[] sourceBlocks = page.getBlocks();
-        Block[] outputBlocks = new Block[sourceBlocks.length + 1]; // +1 for the single boolean output channel
-        System.arraycopy(sourceBlocks, 0, outputBlocks, 0, sourceBlocks.length);
+        Block[] outputBlocks = new Block[page.getChannelCount() + 1]; // +1 for the single boolean output channel
+        for (int channel = 0; channel < page.getChannelCount(); channel++) {
+            outputBlocks[channel] = page.getBlock(channel);
```

Comment:
Here and elsewhere: isn't `System.arraycopy` / `Arrays.copyOf` better?

Reply:
Various places call `Page.getBlocks()` which returns a copy of the array in order to enforce that `Page` is immutable. This can be expensive, but that's not apparent in the API, so we have the mistake in many places.

Note that this usage isn't a pure copy: it's copying the array into a larger array. So, assuming `System.arraycopy` is faster (which the `Arrays` methods use, assuming they aren't JVM intrinsics), we would need a `Page.copyBlocksInto` method.

But I would be interested in seeing a JMH benchmark for the object array copy on the latest Java 9 or Java 10. After inlining, this should be a canonical array copy loop, and I'd be surprised if the JIT does not generate the same code as `System.arraycopy`.

Reply:
Actually, this code is removed in the *"Add Page.appendColumn"* commit

Reply:
I believe in all the places I changed this, it shouldn't matter.  They are all a "once-per-page" operation.

Reply:
> Note that this usage isn't a pure copy: it's copying the array into a larger array. So, assuming System.arraycopy is faster (which the Arrays methods use, assuming they aren't JVM intrinsics), we would need a Page.copyBlocksInto method.

@electrum i think `System.arraycopy` is an intrinsic (`memcpy`?). I also think that JIT may try to capture hand-written copy loops, but I am not sure it always succeeds.
`Array.copyOf` looks much better than hand-written create+fill, but it uses reflection under the hood for array creation.

I have no idea how important is this. Most of this operations aren't performance critical anyway. (as @dain pointed out)